doc/source/reference/arrays.classes.rst

```diff
@@ -39,77 +39,119 @@ Special attributes and methods
+       :func:`~numpy.matmul`, which currently is not a Ufunc, but could be
+       relatively easily be rewritten as a (set of) generalized Ufuncs. The
+       same may happen with functions such as :func:`~numpy.median`,
+       :func:`~numpy.min`, and :func:`~numpy.argsort`.
+
+
+   Like with some other special methods in python, such as ``__hash__`` and
+   ``__iter__``, it is possible to indicate that your class does *not*
+   support ufuncs by setting ``__array_ufunc__ = None``. With this,
+   inside ufuncs, your class will be treated as if it returned
+   :obj:`NotImplemented` (which will lead to an :exc:`TypeError`
+   unless another class also provides a :func:`__array_ufunc__` method
+   which knows what to do with your class).
+
+   The presence of :func:`__array_ufunc__` also influences how
+   :class:`ndarray` handles binary operations like ``arr + obj`` and ``arr
+   < obj`` when ``arr`` is an :class:`ndarray` and ``obj`` is an instance
+   of a custom class. There are two possibilities. If
+   ``obj.__array_ufunc__`` is present and not :obj:`None`, then
+   ``ndarray.__add__`` and friends will delegate to the ufunc machinery,
+   meaning that ``arr + obj`` becomes ``np.add(arr, obj)``, and then
+   :func:`~numpy.add` invokes ``obj.__array_ufunc__``. This is useful if you
+   want to define an object that acts like an array.
+
+   Alternatively, if ``obj.__array_ufunc__`` is set to :obj:`None`, then as a
+   special case, special methods like ``ndarray.__add__`` will notice this
+   and *unconditionally* return :obj:`NotImplemented`, so that Python will
+   dispatch to ``obj.__radd__`` instead. This is useful if you want to define
+   a special object that interacts with arrays via binary operations, but
+   is not itself an array. For example, a units handling system might have
+   an object ``m`` representing the "meters" unit, and want to support the
+   syntax ``arr * m`` to represent that the array has units of "meters", but
+   not want to otherwise interact with arrays via ufuncs or otherwise. This
+   can be done by setting ``__array_ufunc__ = None`` and defining ``__mul__``
+   and ``__rmul__`` methods. (Note that this means that writing an
+   ``__array_ufunc__`` that always returns :obj:`NotImplemented` is not
+   quite the same as setting ``__array_ufunc__ = None``: in the former
+   case, ``arr + obj`` will raise :exc:`TypeError`, while in the latter
+   case it is possible to define a ``__radd__`` method to prevent this.)
+
+   The above does not hold for in-place operators, for which :class:`ndarray`
+   never returns :obj:`NotImplemented`.  Hence, ``arr += obj`` would always
+   lead to a :exc:`TypeError`.  This is because for arrays in-place operations
+   cannot generically be replaced by a simple reverse operation.  (For
+   instance, by default, ``arr[:] += obj`` would be translated to ``arr[:] =
+   arr[:] + obj``, which would likely be wrong.)
```

Comment:
I agree that that's a problem, but it's a problem with python in general, not just numpy:

```
al2 = al = [1, 2, 3]
al += [4, 5]
assert al is al2

at2 = at = (1, 2, 3)
at += (4, 5)
assert at is at2  #AssertionError
```

So the question is - should we be rejecting a questionable design choice made by python because we "know better"? Or should we respect it, just because in some sense users are expecting it.

If this was covered in that long discussion, I apologize.